pane: Apply `max_tabs` change immediately

vipexv opened the pull request:
Closes #32217

Follow up of https://github.com/zed-industries/zed/pull/32301, sorry about the messy rebase in the previous PR.

Release Notes: 
- Fixed `max_tabs` setting not applying immediately when changed
 
TODO: 
- [x] Fix the off-by-one bug (currently closing one more tab than the max_tabs setting) while perserving "+1 Tab Allowance" feature.
- [x] Investigate Double Invocation of `settings_changed`
- [x] Write test that:
  - Sets max_tabs to `n`
  - Opens `n` buffers
  - Changes max_tabs to `n-1`
  - Asserts we have exactly `n-1` buffers remaining

vipexv committed:
pane: Apply `max_tabs` change immediately

vipexv committed:
pane: fix one off issue when closing items over max tabs

vipexv commented:
@JosephTLyons The newest commit should fix the one off bug, and make the behavior the same as VSCode.

https://github.com/user-attachments/assets/5958af15-c38a-4341-a875-51ee46347005

vipexv committed:
pane: revert formatting change when iterating over `index_list`

vipexv committed:
pane: add test - `test_reduce_max_tabs_closes_existing_items`

vipexv commented:
Marked as draft once again, some tests need to be fixed due to the new changes.

vipexv commented:
> > After investigating it some more, i found the root issue for the off-one bug when changing the `max_tabs`, somehow the `observe_global_in` seems to invoke `settings_changed` twice eachtime the settings change leading to a race condition, and i'm unsure why.
> 
> Ahh, that would make sense then as to why we're seeing it kill off too many tabs. That being said, that bit might be too deep in the weeds for me to know of a proper fix and I might need to call on another engineer who knows that area more.

I have a potential workaround for the double-invocation issue, though it's not ideal:

**Option 1: Context-aware logic**
Add a parameter to the tab-closing function that indicates whether it's being called from `settings_changed`. This would let us handle the double-invocation scenario differently.

**Option 2: Adjust the comparison operator**
Change `items_len < max_tabs` to `items_len <= max_tabs`. This fixes the immediate issue, but it breaks the "+1 tab allowance" feature for pinned/dirty tabs that you mentioned.

Both approaches feel like workarounds rather than proper fixes. The root cause is still the double invocation from `observe_global_in`, which ideally should be addressed at the observer level.

Would it be worth investigating why the observer fires twice? Or should we implement one of these workarounds as a temporary solution while the underlying issue gets proper attention from someone more familiar with the observer system?

vipexv commented:
> Having more context as to why `settings_changed` is called twice would be good. I'm curious if there's a real reason for that or if it's a bug. I've asked the team, and I can also try to dig into it a bit myself this week. If it's an unintentional thing, getting that fixed would be awesome. If its intentional and other things rely on it, a workaround could be added, but I'd prefer to avoid losing the +1 tab behavior, as I think blocking a tab when a user needs it is worse than the bug of us not applying `max_tabs` immediately on settings file save.

I'll try and look into it myself when i get the chance, thank you once again!

JosephTLyons commented:
@smitbarmase just hopped into a call with me to show me an existing pattern that might work as a workaround here.

In this code, we observe the settings, but do not run actions on settings change unless the setting value is different from one that was seen last

https://github.com/zed-industries/zed/blob/a9d0eee2a9687c14c5a82d801a185ec280f05cf3/crates/project_panel/src/project_panel.rs#L460-L475

We could do something similar here where we store the `max_tabs` value in the `pane`. 

When observing, we can check if the new setting differs from what we have cached, and avoid double calling `self.close_items_over_max_tabs`.

vipexv committed:
sync with git

vipexv commented:
> @smitbarmase just hopped into a call with me to show me an existing pattern that might work as a workaround here.
> 
> In this code, we observe the settings, but do not run actions on settings change unless the setting value is different from one that was seen last
> 
> https://github.com/zed-industries/zed/blob/a9d0eee2a9687c14c5a82d801a185ec280f05cf3/crates/project_panel/src/project_panel.rs#L460-L475
> 
> We could do something similar here where we store the `max_tabs` value in the `pane`.
> 
> When observing, we can check if the new setting differs from what we have cached, and avoid double calling `self.close_items_over_max_tabs`.

I've just tested the project panel, weird enough theirs doesn't invoke/trigger twice.

JosephTLyons commented:
Interesting! Somehow, a terminal tab is being included as the additional item

```rs
        let item_tabs: Vec<String> = self
            .items
            .iter()
            .map(|item| item.tab_content_text(0, cx).to_string())
            .collect();

        dbg!("Pane::settings_changed", cx.entity_id(), &item_tabs);
```

```
[crates/workspace/src/pane.rs:648:9] "Pane::settings_changed" = "Pane::settings_changed"
[crates/workspace/src/pane.rs:648:9] cx.entity_id() = EntityId(
    79v1,
)
[crates/workspace/src/pane.rs:648:9] &item_tabs = [
    "~/.config/zed/settings.json",
]
[crates/workspace/src/pane.rs:648:9] "Pane::settings_changed" = "Pane::settings_changed"
[crates/workspace/src/pane.rs:648:9] cx.entity_id() = EntityId(
    188v1,
)
[crates/workspace/src/pane.rs:648:9] &item_tabs = [
    "josephlyons — zsh ",
]
```

JosephTLyons committed:
WIP

JosephTLyons commented on crates/workspace/src/pane.rs:
```diff
@@ -1409,29 +1422,59 @@ impl Pane {
         )
     }
 
-    pub fn close_items_over_max_tabs(&mut self, window: &mut Window, cx: &mut Context<Self>) {
-        let Some(max_tabs) = WorkspaceSettings::get_global(cx).max_tabs.map(|i| i.get()) else {
```
In the original code, we always pulled from the settings, but now that we are caching the `max_tabs` value on the `pane` struct, in the updated code, I'm just pulling from that instead inside of `close_tabs_on_file_open` and `close_tabs_on_settings_change`.

JosephTLyons committed:
Rename

vipexv commented:
> Hey @vipexv, I pushed up a commit and added some inline comments and I'm curious what you think. I've also added a bit of extra testing and code to ensure that in the case of closing tabs when the settings file is changed, we dont allow the active file (which is the settings.json file) to be closed, as I think that would likely be a bad experience, so it is now a "protected" item.
> 
> Hope me collaborating with you on this was ok! And if anything seems wrong or off about what I pushed, would love to continue chatting about the next steps.



> Hey @vipexv, I pushed up a commit and added some inline comments and I'm curious what you think. I've also added a bit of extra testing and code to ensure that in the case of closing tabs when the settings file is changed, we dont allow the active file (which is the settings.json file) to be closed, as I think that would likely be a bad experience, so it is now a "protected" item.
> 
> Hope me collaborating with you on this was ok! And if anything seems wrong or off about what I pushed, would love to continue chatting about the next steps.

Thank you for collaborating with me on this! I also appreciate the comments as it helps me and other people understand why it's written that way :], and i also feel like the settings file should always be protected in this case for a better UX.

I'll be looking into this & testing it some more today, thank you once again for taking the time to work on this!

JosephTLyons commented:
Thanks for getting this PR up and going and digging. I went ahead and merged because it felt good to me. We can always forward fix if something else comes up! Love getting all these tests around all the cases that come up with the various tab states and `max_tab` setting mixtures.

Also, nice work on the other PRs you have landed - these fixes make are super appreciated.
